File: clojure/run-length-encoding/src/run_length_encoding.clj
(ns run-length-encoding)

(defn- combine [[consecutive ch]]
	(let [cnt (count consecutive)]
		(if (= cnt 1) ch
			(str cnt ch))))

(defn- expand [[code ch]]
	(let [cnt (count code)]
		(if (= cnt 1) ch 
			(let [repeats (read-string (subs code 0 (dec cnt)))]
				(apply str (repeat repeats ch))))))

(defn run-length-encode [s]
	(if (= s "") ""
		(apply str (map combine (re-seq #"(.)\1{0,}" s)))))

(defn run-length-decode [s]
	(if (= s "") ""
 	 	(apply str (map expand (re-seq #"\d*(.)" s)))))

File: python/run-length-encoding/run_length_encoding.py
import re

def decode(string):
    codes = re.findall("(\d*)(\D)",string)
    return "".join([int(c)*w if c.isdigit() else w for c,w in codes])


def encode(string):
    words = []
    count = []
    for a in string:
        if not words or words[-1] != a:
            words.append(a)
            count.append(1)
        else:
            count[-1] += 1
    return "".join([("" if c == 1 else str(c))+w for w,c in zip(words, count)])
